Colorize summarized RSpec results.

joshuapinter opened the pull request:
RSpec colorizes its results based on if there are any failures (red), pending (orange) or none of the above (green).

The method used to summarize the results from the various threads, `summarize_results` was not doing this colorization.

This commit applies the same colorization that RSpec does, utilizing RSpec::Core helpers to do so.

## Before

![Screenshot 2020-11-08 at 10 27 05](https://user-images.githubusercontent.com/180819/98471962-f6954000-21ac-11eb-9d99-87fe4e5e5bc7.png)

![Screenshot 2020-11-08 at 10 17 20](https://user-images.githubusercontent.com/180819/98471965-fac15d80-21ac-11eb-8dfc-8272d0ea32df.png)

![Screenshot 2020-11-08 at 10 14 45](https://user-images.githubusercontent.com/180819/98471969-fe54e480-21ac-11eb-95e8-1b69bcc622a3.png)


## After

![Screenshot 2020-11-08 at 10 10 25](https://user-images.githubusercontent.com/180819/98471971-03199880-21ad-11eb-82b9-7e22525d8d5c.png)

![Screenshot 2020-11-08 at 10 07 12](https://user-images.githubusercontent.com/180819/98471979-0b71d380-21ad-11eb-974c-4804ba3ecedb.png)

![Screenshot 2020-11-08 at 10 12 26](https://user-images.githubusercontent.com/180819/98471984-10cf1e00-21ad-11eb-9d43-f0c6073e62cb.png)


Thank you for your contribution!

## Checklist
- [x] Feature branch is up-to-date with `master` (if not - rebase it).
- [x] Added tests.
- [x] Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new 
  code introduces user-observable changes.
- [x] Update Readme.md when cli options are changed

grosser commented on lib/parallel_tests/rspec/runner.rb:
```diff
@@ -1,4 +1,5 @@
 require "parallel_tests/test/runner"
+require "rspec"
```
prefer to not depend on rspec (to keep working with old/new versions of the gem) especially for trivial things like color-coding

grosser commented on lib/parallel_tests/rspec/runner.rb:
```diff
@@ -48,6 +49,20 @@ def command_with_seed(cmd, seed)
+        #
+        def summarize_results(results)
+          sums = sum_up_results(results)
+          color = if sums['failure'].positive?
+                    :failure
+                  elsif sums['pending'].positive?
+                    :pending
+                  else
+                    :success
+                  end
+          results_string = sums.sort.map { |word, number| "#{number} #{word}#{'s' if number != 1}" }.join(', ')
```
```suggestion
          text = super
          return text unless $stdout.tty?
          color =
            if sums['failure'].positive?
               31 # red
            elsif sums['pending'].positive?
              33 # yellow
            else
               32 # green
            end
          "\e[#{color}m#{text}\e[0m"
```

grosser replied:
... and even then how about moving this into lib/parallel_tests/test/runner.rb since this is pretty universal ?

grosser replied:
... and there is a good chance some of these keys will be missing, so doing something like `sums.key?('pending')` might be safer

grosser commented:
also: sorry for the long silence, this has been buried in my inbox :D

grosser commented on lib/parallel_tests/rspec/runner.rb:
```diff
@@ -48,6 +48,22 @@ def command_with_seed(cmd, seed)
+        # Summarize results from threads and colorize results based on failure and pending counts.
+        #
+        def summarize_results(results)
+          sums = sum_up_results(results)
```
can move this under the return

grosser commented:
can you add a test for the coloration happening ?

joshuapinter commented:
All good! I honestly forgot about it myself. :)

I took your suggested change and removed the rspec dependency. Added an entry to CHANGELOG as well. 

Will just add some tests and then pass it back to you.

joshuapinter commented:
Quick Q: Are these CI tests supposed to be failing. Seem unrelated to my changes...

grosser commented:
trying to fix CI here https://github.com/grosser/parallel_tests/pull/788

grosser commented:
please rebase, at least ubuntu should be fixed

joshuapinter commented:
Will squash and rebase...

joshuapinter commented:
Odd, it's not letting me rebase and force push. I think because it includes your workflow commit: 

```
 ! [remote rejected] colorize_rspec_results -> colorize_rspec_results (refusing to allow an OAuth App to create or update workflow `.github/workflows/test.yml` without `workflow` scope)
error: failed to push some refs to 'https://github.com/cntral/parallel_tests.git'
```

Super strange. I'm going open a new PR and see if that helps.

grosser commented:
might be some security foobar having to do with the PR being older than the GA addition :(

... can you make a new branch and then pick this commit on top ?

joshuapinter commented:
Damn, I can't even update my own fork's master from your master. 

```
! [remote rejected] master -> master (refusing to allow an OAuth App to create or update workflow `.github/workflows/test.yml` without `workflow` scope)
```

That's restrictive.

I might be forced to create a new fork entirely. Super strange.

joshuapinter commented:
Okay, the only way I could update my fork's `master` was to create a PR from your `master` to my `master` in Github's UI.

Then I was able to rebase my branch and force push. It's still showing your commit in this branch for some reason but 🤷‍♂️ .

Take a look and let me know if you want or need anything else. 

Thanks.

joshuapinter commented:
Kk, nevermind. I was able to fix the extra commit with _another_ rebase. Strange times. Over to you...

grosser commented:
```
       expected: "\e[31m1 example, 1 failure, 1 pending\e[0m"
            got: "1 example, 1 failure, 1 pending"
```
needs an update ?

grosser commented:
... could strip extra characters in all other tests with a regex

grosser commented:
only 1 commit here now :(

joshuapinter commented:
Strange. Okay, fixed that and re-pushed.

joshuapinter commented:
Tests still failing.... All tests pass on macOS. Does Ubuntu somehow deal with these differently?

grosser commented:
might handle tty differently, best to just stub tty? to be true ... or make it accept different results based on `if $stdout.tty?` in the test

joshuapinter commented:
> might handle tty differently, best to just stub tty? to be true ... or make it accept different results based on `if $stdout.tty?` in the test

Good call. 👍

grosser commented:
thx, 3.4.0 🎉

joshuapinter commented:
Thanks for the help and shipping a new version so quickly! 🚢
